src/Task.cc

```diff
@@ -103,28 +103,75 @@ void Task::detach() {
   fallible_ptrace(PTRACE_DETACH, nullptr, nullptr);
 }
 
-Task::~Task() {
-  if (unstable) {
-    LOG(warn) << tid << " is unstable; not blocking on its termination";
-    // This will probably leak a zombie process for rr's lifetime.
+bool Task::may_reap() {
+  // Non thread-group-leaders may always be reaped
+  if (tid != real_tgid())
+    return true;
+  if (thread_group()->task_set().size() > 1)
+    return false;
+  return true;
+}
 
-    // Destroying a Session may result in unstable exits during which
-    // Task::destroy_buffers() will not have been called.
-    if (!syscallbuf_child.is_null()) {
-      vm()->unmap(this, syscallbuf_child, syscallbuf_size);
-    }
+void Task::reap() {
+  ASSERT(this, !was_reaped);
+  LOG(debug) << "Reaping " << tid;
+  siginfo_t info;
+  memset(&info, 0, sizeof(info));
```

Comment:
Why do you zero `info` here but not in `wait_exit`?

Reply:
It needs to be zero'ed if you pass `WNOHANG`, because otherwise you can't distinguish between the WNOHANG case and a successful wait (in the former the kernel doesn't touch the info struct). In `wait_exit`, we don't pass `WNOHANG`.